core/Segment/SegmentExpression.php

```diff
@@ -456,47 +459,25 @@ private function escapeLikeString($str)
      */
     protected function parseTree()
     {
-        $string = $this->string;
-        if (empty($string)) {
-            return array();
+        $segmentStr = trim($this->string);
+        if (empty($segmentStr)) {
+            return [];
         }
-        $tree = array();
-        $i = 0;
-        $length = strlen($string);
-        $isBackslash = false;
-        $operand = '';
-        while ($i <= $length) {
-            $char = $string[$i];
-
-            $isAND = ($char == self::AND_DELIMITER);
-            $isOR = ($char == self::OR_DELIMITER);
-            $isEnd = ($length == $i + 1);
-
-            if ($isEnd) {
-                if ($isBackslash && ($isAND || $isOR)) {
-                    $operand = substr($operand, 0, -1);
-                }
-                $operand .= $char;
-                $tree[] = array(self::INDEX_BOOL_OPERATOR => self::BOOL_OPERATOR_END, self::INDEX_OPERAND => $operand);
-                break;
-            }
 
-            if ($isAND && !$isBackslash) {
-                $tree[] = array(self::INDEX_BOOL_OPERATOR => self::BOOL_OPERATOR_AND, self::INDEX_OPERAND => $operand);
-                $operand = '';
-            } elseif ($isOR && !$isBackslash) {
-                $tree[] = array(self::INDEX_BOOL_OPERATOR => self::BOOL_OPERATOR_OR, self::INDEX_OPERAND => $operand);
-                $operand = '';
-            } else {
-                if ($isBackslash && ($isAND || $isOR)) {
-                    $operand = substr($operand, 0, -1);
-                }
-                $operand .= $char;
-            }
-            $isBackslash = ($char == "\\");
-            $i++;
-        }
-        return $tree;
+        // handle backslashes by urlencoding the characters
+        $segmentStr = preg_replace_callback('/\\\\./', function ($matches) {
+            $char = substr($matches[0], 1);
+            return urlencode($char);
+        }, $segmentStr);
```

Comment:
We might need to revert all changes to this method. We can't simply urlencode all values that follow on a `\`.
There might e.g. be segment values where someone wants to look after `folder\file`. The current solution would simply strip away the `\`. Also we can not start to urlencode various chars, as they might then appear urlencoded in the resulting SQL (which can be seen in the test changes), which could have an effect on the resulting datasets.

Here a small php example what the problem around backslashes are: https://3v4l.org/8kpMn

Reply:
@sgiehl Unfortunately I couldn't just revert the `parseTree()` method since as well as changing the handling of backslashes it also changes the grouping of the return array. Instead I've reworked the method to use the original parsing code but tweaked to provide the new grouped array. As part of this I've reverted the test values in `SegmentExpressionTest` back to their original state which includes the use of a backslash as a non-escaping part of the expression (`A\B`).

Both `SegmentExpressionTest` and the recently updated `ComplicatedSegmentTest` are now passing.

Reply:
Had a quick play and this may also produce the same nested tree.
It's using a negative lookbehind split, first splits on `;` that is not preceded by `\` and then splits the results on that on `,` that is not preceded by `\`. Then removes `\` from `\;` and `\,`.

```
function parseTree($segment) {
    $ands = preg_split("/;(?<!\\\;)/", $segment);
    $tree = array_map(function($and) {
        $ors = preg_split("/,(?<!\\\,)/", $and);
        $unescapedOrs = array_map(function($or) { return str_replace(['\,', '\;'], [',', ';'], $or); }, $ors);
        return $unescapedOrs;
    }, $ands);

    return $tree;
}
```
(would need replacing `;` with `self::AND_DELIMITER` and `,` with `self::OR_DELIMITER` in the preg_split expressions).

Reply:
Mmmm, would need more tweaks, run the unit test locally and failed on some trailing commas.

Reply:
I got the tests passing locally, have a PR on top of this PR to test it.

Reply:
@bx80 check https://github.com/matomo-org/matomo/pull/21669 and see if it works for you.

Reply:
That change now passes the tests, but there may be other edge cases not covered by the tests so we're still introducing additional risk.

Reply:
Yep, fair. We can leave it open even after this is merged, it will automatically retarget itself to 5.x-dev and we can review/add extra tests then.